### ISSUE TITLE
What's the shape of the result

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1249]
**Reported by ansley b manke on 18 Oct 2012 00:26 UTC**
This query on the Users List,
http://www.pmel.noaa.gov/maillists/tmap/ferret_users/fu_2012/msg00569.html

wonders what would it take to "shape" of an expression without loading the data and making computations. Some discussion here. This would be nice to do sometime but at low priority.

Note that the task in ticket #1976 puts in place at least a nice way to get out of an over-long data read.

Ferret includes a tremendous amount of code -- the entire "get_grid" and context processing support phases of evaluation -- that exist for the express purpose of figuring out the size and shape of a result *without* ever loading data.  In the example session pasted below, the MODE DIAG output shows that no data are ever read, yet the size and shape of the result can be seen.  The final command, shade sst" illustrates a case where the entire domain of data is read, only to result in an error condition over shape.  Ryo's right.  This is not optimal behavior.

The trouble lies in the exceptional cases (compressing transformations like @AVE btw, are handled fine).  There are flaws that keep the pre-determining of size/shape from working as it should in all cases -- a collection of specific conditions that force Ferret to read data in order to know the shape of a result.  I'm feeling pretty fuzzy on a lot of those details at this point (it's been a while), but some of the specific things that inhibit pre-determining the size of a result include

    reading from ASCII files, where the file must be read to know how long it really is
    grid-changing functions where the output grid definition depend upon the input *data* values
    some cases of limits being implicit instead of explicit ...

Bottom line:  Ferret *does* have the inner machinery capable of doing exactly what Ryo has requested **most of the time**.   And it would be really a nice addition to do so -- he's right.  Its just that the exceptional cases (which I do not remember so well), can turn into bad behaviors -- sometimes making Ferret do a big data access task twice, sometimes leading to an error condition.

It would be an interesting prospect to add the check that Ryo speaks of in some kind of conditional, experimental manner:

    an internal list of commands that have shape restrictions -- e.g. PLOT required 1D args, SHADE requires 2D, ... 
    some new code that would be invoked by those commands to pre-screen the data arguments for correct shape (without reading any data) and quickly trigger error conditions if so

    ... but probably also needing these to cope with the exceptional cases ...

    a MODE CHECK_SHAPE (so you can turn off the attempts at shape pre-screening)
    qualifier /OVERRIDE_SHAPE to bypass this where it leads to trouble (YUK! UGLY)

However, I don't think this issue rises high enough in the priority list to actually do the work.  It only saves users from cases of their own carelessness -- it doesn't add new power.  But there are hundreds of times I would have enjoyed the feature myself!  if it were reliable, that is ...

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1977

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 20 Dec 2012 23:52 UTC**
I too would like to see this feature for the reasons described above.  But beyond that, I'd also argue that this *would* add new power, since it would allow scripts to check whether their arguments have the proper shape to be processed.  Right now, we have weirdness like this:

```
 	NOAA/PMEL TMAP
 	FERRET v6.83 (beta)
 	Linux 2.6.32-279.5.1.el6.x86_64 64-bit - 08/22/12
 	20-Dec-12 18:44     

yes? let a = {1,2,3}+ysequence({4,5,6})
yes? let b = a[x=@ave]
yes? say `a,r=shape`
 !-> MESSAGE/CONTINUE XY
XY
```

which is clearly wrong, the answer should really be "Y".  Well, maybe there's a "compute-domain shape" XY and a "result shape" Y, but we really need a way to get at the latter.  Right now I often have scripts dump the darn variable to a NetCDF file and read it back in, just to get its shape - yuck!

So I'm bumping up the priority on this...

---

EDIT: OOPS! -- You're right Steve that I goofed on my example.  (If I had pasted "b" in there instead of "a" I would've seen that my example didn't demonstrate the issue at all -- sorry for confusing everyone!)  Thanks Ansley for correctly highlighting the real issue, which must have to do with source data that needs to be LOADed.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 21 Dec 2012 00:08 UTC**
Puzzled by this.  In the above the shape of "a" is XY, isn't it?  It is variable "b", where the X axis has been compressed out and for that we get  
```
yes?  yes? say `b,r=shape`
 !-> MESSAGE/CONTINUE Y
```
Having said this, I would not be surprised to find examples (pyramids of definitions), where the true shape was getting lost.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 21 Dec 2012 00:15 UTC**
The example in comment 1 doesn't capture the issue.  for those variables the data is loaded and then the shape is returned.  But,

yes? use coads_climatology
yes? say `sst,return=shape`
 !-> MESSAGE/CONTINUE XYT

doesn't load the data. It does a "getgrid" operation and returns the correct answer. But the shape of this variable is really XY and Ferret doesn't go far enough evaluating the grid to get that:

yes? say `sst[L=@max],return=shape`
 !-> MESSAGE/CONTINUE XYT

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 26 Dec 2012 22:34 UTC**
Specifically, this issue applies to transformations with indefinite limits.  See ticket #1801.

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 6 Jan 2013 17:57 UTC**
This ticket articulated 3 cases where it would be desirable to avoid reading the data in order to determine the shape and size of the result:

   1. reading from ASCII files, where the file must be read to know how long it really is
   2. grid-changing functions where the output grid definition depend upon the input *data* values
   3. some cases of limits being implicit instead of explicit ...
The first two cases require reading the data by definition.  There is no general route around them.  The only articulated example of the third case has now been fixed in ticket #1801.
